lib/class-wp-rest-url-details-controller.php

```diff
@@ -0,0 +1,246 @@
+			'title'      => 'url-details',
+			'type'       => 'object',
+			'properties' => array(
+				'title' => array(
+					'description' => __( 'The contents of the <title> tag from the URL.', 'gutenberg' ),
+					'type'        => 'string',
+					'default'     => '',
+					'context'     => array( 'view', 'edit', 'embed' ),
+					'readonly'    => true,
+				),
+			),
+		);
+	}
+
+	/**
+	 * Retrieves the contents of the <title> tag from the HTML
+	 * response.
+	 *
+	 * @access public
+	 * @param WP_REST_REQUEST $request Full details about the request.
+	 * @return String|WP_Error The title text or an error.
+	 */
+	public function parse_url_details( $request ) {
+
+		$url = untrailingslashit( $request->get_param( 'url' ) );
+
+		if ( empty( $url ) ) {
+			return new WP_Error( 'rest_invalid_url', __( 'Invalid URL', 'gutenberg' ), array( 'status' => 404 ) );
+		}
+
+		// Transient per URL.
+		$cache_key = $this->build_cache_key_for_url( $url );
+
+		// Attempt to retrieve cached response.
+		$data = $this->get_cache( $cache_key );
+
+		$use_cache = true;
+
+		/**
+		 * Filters whether to check for previously cached
+		 * data for a given URL. Filtering for `false` will
+		 * mean that the cache is bypassed.
+		 *
+		 * @param bool $use_cache whether or not to use the cache
+		 * @param string $url the attempted URL.
+		 * @param string $cache_key the generated cache key for the given URL.
+		 */
+		if ( apply_filters( 'rest_url_details_use_cache', $use_cache, $url, $cache_key ) && ! is_wp_error( $data ) && ! empty( $data ) ) {
+			return json_decode( $data, true );
+		}
+
+		$response_body = $this->get_remote_url( $url );
+
+		// Exit if we don't have a valid body or it's empty.
+		if ( is_wp_error( $response_body ) || empty( $response_body ) ) {
+			return $response_body;
+		}
+
+		$data = array(
+			'title' => $this->get_title( $response_body ),
+		);
+
+		// Cache the valid response.
+		$this->set_cache( $cache_key, $data );
+
+		return rest_ensure_response( $data );
+	}
+
+	/**
+	 * Checks whether a given request has permission to read remote urls.
+	 *
+	 * @return WP_Error|bool True if the request has access, WP_Error object otherwise.
+	 */
+	public function permissions_check() {
+		if ( current_user_can( 'edit_posts' ) ) {
+			return true;
+		}
+
+		foreach ( get_post_types( array( 'show_in_rest' => true ), 'objects' ) as $post_type ) {
+			if ( current_user_can( $post_type->cap->edit_posts ) ) {
+				return true;
+			}
+		}
+
+		return new WP_Error(
+			'rest_user_cannot_view',
+			__( 'Sorry, you are not allowed to process remote urls.', 'gutenberg' ),
+			array( 'status' => rest_authorization_required_code() )
+		);
+	}
+
+	/**
+	 * Retrieves the document title from a remote URL.
+	 *
+	 * @param string $url The website url whose HTML we want to access.
+	 * @return string|WP_Error The URL's document title on success, WP_Error on failure.
+	 */
+	private function get_remote_url( $url ) {
+
+		$args = array(
+			'timeout'             => 10,
```

Comment:
Why isn't the default of 5 enough?

Reply:
I actually dont know. 5 should be plenty right? If it takes longer we dont want to wait around.

Reply:
Yes I think so too!

Reply:
Fixed